include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -34,332 +35,162 @@ namespace beast {
-{
-    return const_iterator{*b_, b_->begin_};
-}
-
-template<class MutableBufferSequence>
-template<bool isMutable>
-auto
-buffers_adaptor<MutableBufferSequence>::
-readable_bytes<isMutable>::
-readable_bytes::end() const ->
-    const_iterator
-{
-    return const_iterator{*b_, b_->end_impl()};
-}
+#if BOOST_WORKAROUND(BOOST_MSVC, < 1910)
+# pragma warning (pop)
+#endif
 
 //------------------------------------------------------------------------------
 
 template<class MutableBufferSequence>
-class buffers_adaptor<MutableBufferSequence>::
-mutable_buffers_type
+template<bool isMutable>
+struct buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator
 {
-    buffers_adaptor const* b_;
-
-public:
-    using value_type = net::mutable_buffer;
-
-    class const_iterator;
-
-    mutable_buffers_type() = delete;
-    mutable_buffers_type(
-        mutable_buffers_type const&) = default;
-    mutable_buffers_type& operator=(
-        mutable_buffers_type const&) = default;
-
-    const_iterator
-    begin() const;
-
-    const_iterator
-    end() const;
-
-private:
-    friend class buffers_adaptor;
-
-    mutable_buffers_type(
-            buffers_adaptor const& b)
-        : b_(&b)
+    struct iterator_category
+        : std::bidirectional_iterator_tag
     {
-    }
-};
-
-template<class MutableBufferSequence>
-class buffers_adaptor<MutableBufferSequence>::
-mutable_buffers_type::const_iterator
-{
-    iter_type it_{};
-    buffers_adaptor const* b_ = nullptr;
-
-public:
-    using value_type = net::mutable_buffer;
-    using pointer = value_type const*;
-    using reference = value_type;
+    };
+    using value_type = typename buffers_adaptor<MutableBufferSequence>::
+    template subrange<isMutable>::
+    value_type;
+    using reference = typename std::add_lvalue_reference<value_type>::type;
+    using pointer = typename std::add_pointer<value_type>::type;
     using difference_type = std::ptrdiff_t;
-    using iterator_category =
-        std::bidirectional_iterator_tag;
 
-    const_iterator() = default;
-    const_iterator(const_iterator const& other) = default;
-    const_iterator& operator=(const_iterator const& other) = default;
+    iterator(
+        typename buffers_adaptor<MutableBufferSequence>::
+        template subrange<isMutable> const *parent,
+        iter_type it);
 
-    bool
-    operator==(const_iterator const& other) const
-    {
-        return b_ == other.b_ && it_ == other.it_;
-    }
+    iterator();
 
-    bool
-    operator!=(const_iterator const& other) const
-    {
-        return !(*this == other);
-    }
+    value_type
+    operator*() const;
 
-    reference
-    operator*() const
-    {
-        value_type const b = *it_;
-        return value_type{b.data(),
-            it_ == std::prev(b_->end_) ?
-                b_->out_end_ : b.size()} +
-                    (it_ == b_->out_ ? b_->out_pos_ : 0);
-    }
+    iterator &
+    operator++();
 
-    pointer
-    operator->() const = delete;
```

Comment:
Why did you remove this?

Reply:
the iterator is now a subtype of subrange. It wasn't that I deleted this specifically, I just didn't write it. What is the purpose of this deleted function?

Reply:
to inform the user that the omission is intended (`operator->` is unimplementable)